### PR TITLE
Assignment 3 - Zach Lawson and Tom Rohr

### PR DESCRIPTION
**Pull request by zlawson-ewu**
Assignment 3 pair programed by Zach Lawson (@zlawson-ewu) and Tom Rohr (@Tom-Rohr). Both extra credit items completed: JokeService.GetJoke() pulls joke in JSON form and returns it as a string, and the console writer class is unit tested.

### CODE REVIEW

**Comment by arose-13** [comment]
I'm reviewing this

File: CanHazFunny/CanHazFunny.Tests/JesterTests.cs [hunk 1]
```diff
@@ -1,10 +1,82 @@
...
+        //Act
+        new Jester(jokeService, null!);
+    }
+
+    [TestMethod]
+    public void Jester_TellJoke_SkipsChuckNorris()
+    {
+        //Arrange
+        var mockService = new Mock<IJokeService>();
+        mockService.SetupSequence(x => x.GetJoke())
```
Comment (arose-13) [inline CanHazFunny/CanHazFunny.Tests/JesterTests.cs:65]: I would have created a class within this test file that is used as a mock IJokeService (similar to `TestLogger` in assignment 2), but this works too!

**Review by arose-13 — COMMENTED** [review]
- Create *two* interfaces. One to represent the `JokeService` and one interface to represent outputting the joke to the screen (`Console.WriteLine`). ✔
- The `JokeService` will need to have the interface applied to it. ✔
- Create an implementation of the output interface that writes out the joke to the console. ✔
- Implement the `Jester` class. It should take in both interfaces as dependencies. These dependencies should be null checked. ✔
- The `Jester` class `TellJoke()` method should retrieve a joke from the `JokeService`. If the joke contains "Chuck Norris", skip it and get another. The joke should be written to the output dependency. ✔
- Unit test the Jester class. Code coverage should be above 90% for this class. Moq will make this significantly easier. ✔

## Fundamentals

- Be sure you enable:
  - Nullability for all projects  ✔
  - Set `LangVersion` and the `TargetFramework` to the latest released versions available (preview versions optional) ✔
  - Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
- For this assignment, always use `Assert.AreEqual<T>()` (the generic version)  ✔
- **Ensure there are no errors or warnings (including code analysis warnings)** ✔
- **All of the above should be unit tested** ✔
- **Choose simplicity over complexity** ✔
  
## Extra Credit

- Unit test your implementation that writes the joke out to the screen. How hard could it be to unit test a single line method ;)? ✔
- The Geek jokes API that is being used can also return jokes in a JSON format. Update the `JokeService` to retrieve jokes using JSON. `GetJoke` should still return a string. ✔

Well done!

File: CanHazFunny/CanHazFunny.Tests/JesterTests.cs (hunk 1 above)
Comment (zlawson-ewu) [reply CanHazFunny/CanHazFunny.Tests/JesterTests.cs:65]: We did think about that but once we noticed that moq can track method calls, this way seemed much simpler. Moq's documentation really needs some work, though.

**Comment by JessicaH46** [comment]
I am reviewing this

**Comment by JessicaH46** [comment]
Create two interfaces. One to represent the JokeService and one interface to represent outputting the joke to the screen (Console.WriteLine). ✔
The JokeService will need to have the interface applied to it. ✔
Create an implementation of the output interface that writes out the joke to the console. ✔
Implement the Jester class. It should take in both interfaces as dependencies. These dependencies should be null checked. ✔
The Jester class TellJoke() method should retrieve a joke from the JokeService. If the joke contains "Chuck Norris", skip it and get another. The joke should be written to the output dependency. ✔
Unit test the Jester class. Code coverage should be above 90% for this class. Moq will make this significantly easier. ✔
Fundamentals
Be sure you enable:
Nullability for all projects ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
For this assignment, always use Assert.AreEqual<T>() (the generic version) ✔
Ensure there are no errors or warnings (including code analysis warnings) ✔
All of the above should be unit tested ✔
Choose simplicity over complexity ✔

Extra Credit
Unit test your implementation that writes the joke out to the screen. How hard could it be to unit test a single line method ;)?
The Geek jokes API that is being used can also return jokes in a JSON format. Update the JokeService to retrieve jokes using JSON. GetJoke should still return a string.

All the main code looks good and the extra credit also looks good from what I can tell, I didn't attempt the EC but when I ran yours it worked so, great job!

**Pull request closed without merging**